frame/dvm/src/lib.rs

```diff
@@ -574,6 +511,117 @@ impl<T: Config> Pallet<T> {
 			}
 		}
 	}
+
+	/// Transfer rpc transaction to dvm transaction
+	fn to_dvm_transaction(
+		transaction: ethereum::Transaction,
+	) -> Result<DVMTransaction, DispatchError> {
+		let source = recover_signer(&transaction).ok_or(Error::<T>::InvalidSignature)?;
+		Ok(DVMTransaction {
+			source,
+			gas_price: Some(transaction.gas_price),
+			tx: transaction,
+			is_internal: false,
+		})
+	}
+
+	/// Save ethereum block
+	fn store_block(post_log: bool, block_number: U256) {
+		let mut transactions = Vec::new();
+		let mut statuses = Vec::new();
+		let mut receipts = Vec::new();
+		let mut logs_bloom = Bloom::default();
+		for (transaction, status, receipt) in Pending::<T>::get() {
+			transactions.push(transaction);
+			statuses.push(status);
+			receipts.push(receipt.clone());
+			Self::logs_bloom(receipt.logs.clone(), &mut logs_bloom);
+		}
+
+		let ommers = Vec::<ethereum::Header>::new();
+		let partial_header = ethereum::PartialHeader {
+			parent_hash: Self::current_block_hash().unwrap_or_default(),
+			beneficiary: <Pallet<T>>::find_author(),
+			state_root: T::StateRoot::get(),
+			receipts_root: H256::from_slice(
+				Keccak256::digest(&rlp::encode_list(&receipts)[..]).as_slice(),
+			), // TODO: check receipts hash.
+			logs_bloom,
+			difficulty: U256::zero(),
+			number: block_number,
+			gas_limit: T::BlockGasLimit::get(),
+			gas_used: receipts
+				.clone()
+				.into_iter()
+				.fold(U256::zero(), |acc, r| acc + r.used_gas),
+			timestamp: UniqueSaturatedInto::<u64>::unique_saturated_into(
+				<pallet_timestamp::Pallet<T>>::get(),
+			),
+			extra_data: Vec::new(),
+			mix_hash: H256::default(),
+			nonce: H64::default(),
+		};
+		let block = ethereum::Block::new(partial_header, transactions, ommers);
+
+		CurrentBlock::<T>::put(block.clone());
+		CurrentReceipts::<T>::put(receipts);
+		CurrentTransactionStatuses::<T>::put(statuses);
+
+		if post_log {
+			let digest = DigestItem::<T::Hash>::Consensus(
+				FRONTIER_ENGINE_ID,
+				PostLog::Hashes(dp_consensus::Hashes::from_block(block)).encode(),
+			);
+			<frame_system::Pallet<T>>::deposit_log(digest);
+		}
+	}
+
+	fn logs_bloom(logs: Vec<Log>, bloom: &mut Bloom) {
+		for log in logs {
+			bloom.accrue(BloomInput::Raw(&log.address[..]));
+			for topic in log.topics {
+				bloom.accrue(BloomInput::Raw(&topic[..]));
+			}
+		}
+	}
+
+	/// The dispatch result depends on the transaction type.
+	/// The internal transaction will return DispatchError when exitReason does not Succeed,
+	/// The Rpc transaction will alway return Ok(...) no matter what exitReason is.
+	pub fn return_dispatch_result(
```

Comment:
Could we just directly return.reason in this method, and match and give different result according to the call context.

Reply:
No, we can't. This method return `DispatchResultWithPostInfo` and no concrete infomation.

```rs
pub type DispatchResultWithPostInfo = sp_runtime::DispatchResultWithInfo<crate::weights::PostDispatchInfo>;

pub struct PostDispatchInfo {
	/// Actual weight consumed by a call or `None` which stands for the worst case static weight.
	pub actual_weight: Option<Weight>,
	/// Whether this transaction should pay fees when all is said and done.
	pub pays_fee: Pays,
}
```